src/libraries/System.Private.CoreLib/src/System/Double.cs

```diff
@@ -188,6 +188,7 @@ public int CompareTo(object? value)
             throw new ArgumentException(SR.Arg_MustBeDouble);
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
How common are NaNs in the scenarios involving CompareTo?  If we're going to increase call sites for this (including generic specialization), I'm wondering if we should push out the extra asm for handling NaNs into a separate method, e.g.
```C#
    public int CompareTo(double value)
    {
        if (m_value < value) return -1;
        if (m_value > value) return 1;
        if (m_value == value) return 0;
        return CompareToWithAtLeastOneNan(value);
    }
    
    private int CompareToWithAtLeastOneNan(double value)
    {
        Debug.Assert(double.IsNaN(m_value) || double.IsNaN(value));
        if (double.IsNaN(m_value))
            return double.IsNaN(value) ? 0 : -1;
        else
            return 1;
    }
```

Reply:
@stephentoub, it depends on scenario. `NaN` is commonly used as a marker value (including in scenarios such as WPF or ML.NET).

This is really a question of inlinability and so provided the JIT can inline the base method, particularly where constants are inlined, I think we'll be fine. In particular, the JIT kind-of "gives up" on certain optimizations when it knows it has a `NaN` and it could have better awareness of it. Likewise, this entire method could be intrinsified to effectively a single `ucomiss` instruction + the relevant branches (or even `cmovcc`) which would likewise greatly reduce the overall cost.

Reply:
For my code, `NaN` generally indicates an error -- usually a failure to deal with the not-so-subtleties of floating point precision, or not checking for `NaN` earlier in a series of computations.

For the ML / marker stuff, does it also use `CompareTo`? If sorting a list of values that contain `NaN`s, they should all slide to the end of the sorting order (if I'm reading the code correctly).

I do really like the idea of reimplementing this using intrinsics to avoid the branching altogether and also improve its inlining eligibility. Is that possible with the current state of .NET intrinsics?

Reply:
@rickbrew the cutoff for .NET 6.0 is Aug 17th. If the code written like above in two methods would be a strict improvement over today, my suggestion would be to do that.

Reply:
Oh, I see there's discussion of that below. But, the date applies.